File: library-domain/src/main/java/com/dodecaedro/library/domain/repository/BorrowRepository.java
package com.dodecaedro.library.domain.repository;

import com.dodecaedro.library.domain.model.Book;
import com.dodecaedro.library.domain.model.Borrow;
import com.dodecaedro.library.domain.model.User;

import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;

public interface BorrowRepository {
	Borrow save(Borrow borrow);

	Optional<Borrow> findByBookAndUser(Book book, User user);

	List<Borrow> findAll();

	int countExpiredBorrows(User user, ZonedDateTime nowDate);

	int countActiveBorrows(User user);
}

File: library-domain/src/main/kotlin/com/dodecaedro/library/domain/repository/BorrowRepository.kt
package com.dodecaedro.library.domain.repository

import com.dodecaedro.library.domain.model.Book
import com.dodecaedro.library.domain.model.Borrow
import com.dodecaedro.library.domain.model.User
import java.time.ZonedDateTime
import java.util.*

interface BorrowRepository {
	fun save(borrow: Borrow): Borrow

	fun findByBookAndUser(book: Book, user: User): Borrow?

	fun findAll(): List<Borrow>

	fun countExpiredBorrows(user: User, nowDate: ZonedDateTime): Int

	fun findAllExpiredBorrows(time: ZonedDateTime): List<Borrow>

	fun countActiveBorrows(user: User): Int
}
